compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1107,6 +1110,28 @@ trait Applications extends Compatibility {
       tree
   }
 
+  /** If `tree` is a complete application of a compiler-generated `apply`
+   *  or `copy` method of an enum case, widen its type to the underlying
+   *  type by means of a type ascription, as long as the widened type is
+   *  still compatible with the expected type.
+   *  The underlying type is the intersection of all class parents of the
+   *  orginal type.
+   */
+  def widenEnumCase(tree: Tree, pt: Type)(using Context): Tree =
+    val sym = tree.symbol
+    def isEnumCopy = sym.name == nme.copy && sym.owner.isEnumCase
+    def isEnumApply = sym.name == nme.apply && sym.owner.linkedClass.isEnumCase
+    if sym.is(Synthetic) && (isEnumApply || isEnumCopy)
+       && tree.tpe.classSymbol.isEnumCase
+       && tree.tpe.widen.isValueType
```

Comment:
Why is this check needed?

Reply:
So that we don't widen partially applied applications.